### PR TITLE
Optimize join when build side is unique

### PR DESCRIPTION
**Pull request by skrzypo987**
Rebased on top of https://github.com/trinodb/trino/pull/13352.
Only last two commits are relevant

## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

improvement
> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

core query engine
> How would you describe this change to a non-technical end user or system administrator?

Increase performance of join when build side is unique
## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Section
* Increase performance of join when build side is unique
```

### CODE REVIEW

**Comment by findepi** [comment]
> > How would you describe this change to a non-technical end user or system administrator?
> 
> Increase performance of join when build side is unique

How is it achieved?

**Comment by skrzypo987** [comment]
> > > How would you describe this change to a non-technical end user or system administrator?
> > 
> > 
> > Increase performance of join when build side is unique
> 
> How is it achieved?

Are you asking because you're curious or I should rephrase the description?
It is rather technical and includes words like hard-coding which non-technical people do not like.

**Comment by findepi** [comment]
> Are you asking because you're curious or I should rephrase the description?

both

> It is rather technical and includes words like hard-coding which non-technical people do not like.

Fortunately the primary objective of a PR description is to be informative to technical people.

**Comment by skrzypo987** [comment]
The part that you asked about is exactly about "non-technical end user".

I will add some technical description once batching lands and this PR can actually be merged

**Comment by sopel39** [comment]
@skrzypo987 do you have benchmark numbers ?

**Comment by sopel39** [comment]
@skrzypo987 could you also rebase?

**Comment by skrzypo987** [comment]
I ran benchmarks after the rebase and the results are somewhere between no change and a slight regression.
So in this shape this PR does not improve anything. I will post/rebase/benchmark the RLE changes later today and we'll see if it helps.

File: core/trino-main/src/main/java/io/trino/operator/join/DefaultPagesHash.java
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/operator/join/DefaultPagesHash.java:263]: not sure if this is part of this PR but would be good to have this available in OperatorStats

**Review by lukasz-stec — APPROVED** [review]
lgtm

**Comment by lukasz-stec** [comment]
> I ran benchmarks after the rebase and the results are somewhere between no change and a slight regression. So in this shape this PR does not improve anything. I will post/rebase/benchmark the RLE changes later today and we'll see if it helps.

IMO it would beneficial to merge this if we have JMH benchmarks that show improvement here, even if tpch/tpcds do not confirm that as this can be a good step that opens up further optimizations.

**Comment by skrzypo987** [comment]
> > I ran benchmarks after the rebase and the results are somewhere between no change and a slight regression. So in this shape this PR does not improve anything. I will post/rebase/benchmark the RLE changes later today and we'll see if it helps.
> 
> IMO it would beneficial to merge this if we have JMH benchmarks that show improvement here, even if tpch/tpcds do not confirm that as this can be a good step that opens up further optimizations.

I don't think I agree with you. Those changes may be neutral at that point and there is no reason to increase complexity, unless we actually see any reasonable benchmark results.

I will run benchmarks again after merging https://github.com/trinodb/trino/pull/14493. Maybe it will help

One way or another this PR may be a base for some more join optimisations, given someone is going to take over.

**Comment by mosabua** [comment]
@skrzypo987 @lukasz-stec is this still in progress or should we close this PR?

**Comment by lukasz-stec** [comment]
@mosabua this is potentially valuable but no one is working on this AFAIK. If this means we close, lets close.

**Pull request closed without merging**
